plugins/catalog-backend-module-aws/src/providers/AwsS3EntityProvider.ts

```diff
@@ -0,0 +1,208 @@
+    return async () => {
+      const taskId = `${this.getProviderName()}:refresh`;
+      return schedule.run({
+        id: taskId,
+        fn: async () => {
+          const logger = this.logger.child({
+            class: AwsS3EntityProvider.prototype.constructor.name,
+            taskId,
+            taskInstanceId: uuid.v4(),
+          });
+
+          try {
+            await this.refresh(logger);
+          } catch (error) {
+            logger.error(error);
+          }
+        },
+      });
+    };
+  }
+
+  /** {@inheritdoc @backstage/plugin-catalog-backend#EntityProvider.getProviderName} */
+  getProviderName(): string {
+    return `awsS3-provider:${this.config.id}`;
+  }
+
+  /** {@inheritdoc @backstage/plugin-catalog-backend#EntityProvider.connect} */
+  async connect(connection: EntityProviderConnection): Promise<void> {
+    this.connection = connection;
+    await this.scheduleFn();
+  }
+
+  async refresh(logger: Logger) {
+    if (!this.connection) {
+      throw new Error('Not initialized');
+    }
+
+    logger.info('Discovering AWS S3 objects');
+
+    const keys = await this.listAllObjectKeys();
+    logger.info(`Discovered ${keys.length} AWS S3 objects`);
+
+    const locations = keys.map(key => this.createLocationSpec(key));
+
+    await this.connection.applyMutation({
+      type: 'full',
+      entities: locations.map(location => {
+        return {
+          locationKey: this.getProviderName(),
+          entity: locationSpecToLocationEntity({ location }),
+        };
+      }),
+    });
+
+    logger.info(`Committed ${locations.length} Locations for AWS S3 objects`);
+  }
+
+  private async listAllObjectKeys(): Promise<string[]> {
+    const keys: string[] = [];
+
+    let continuationToken: string | undefined = undefined;
+    let output: ListObjectsV2Output;
+    do {
+      const request = this.s3.listObjectsV2({
+        Bucket: this.config.bucketName,
+        ContinuationToken: continuationToken,
+        Prefix: this.config.prefix,
+      });
+
+      output = await request.promise();
+      if (output.Contents) {
+        output.Contents.forEach(item => {
+          if (item.Key && !item.Key.endsWith('/')) {
+            keys.push(item.Key);
+          }
+        });
+      }
+      continuationToken = output.NextContinuationToken;
+    } while (continuationToken);
+
+    return keys;
+  }
+
+  private createLocationSpec(key: string): LocationSpec {
+    return {
+      type: 'url',
+      target: this.createObjectUrl(key),
+      presence: 'required',
+    };
+  }
+
+  private createObjectUrl(key: string): string {
+    const bucketName = this.config.bucketName;
+    const endpoint = this.integration.config.endpoint;
+
+    if (endpoint) {
+      if (endpoint.startsWith(`https://${bucketName}.`)) {
+        return `${endpoint}/${key}`;
+      }
+
+      return `${endpoint}/${bucketName}/${key}`;
+    }
+
+    return `https://${bucketName}.s3.${this.config.region}.amazonaws.com/${key}`;
```

Comment:
we would have to check if the target was really an amazon one right?

Reply:
We already check these cases in the lines before. If we don't point to the real AWS S3, we need to provide the `endpoint` config property (e.g., when LocalStack is used, etc.). So, I'd say it is already covered.